index.js

```diff
@@ -359,6 +406,50 @@ module.exports = function AsyncAppendOnlyLog(filename, opts) {
     })
   }
 
+  function overwrite(blockIndex, blockBuf, cb) {
+    cache.set(blockIndex, blockBuf)
+    const blockStart = blockIndex * blockSize
+    writeWithFSync(blockStart, blockBuf, null, cb)
+  }
+
+  function truncate(newLatestBlockIndex, cb) {
+    if (newLatestBlockIndex >= latestBlockIndex) return cb(null, 0)
+    const size = (latestBlockIndex + 1) * blockSize
+    const newSize = (newLatestBlockIndex + 1) * blockSize
+    for (let i = newLatestBlockIndex + 1; i < latestBlockIndex; ++i) {
+      cache.remove(i)
+    }
+    truncateWithFSync(newSize, function onTruncateWithFSyncDone(err) {
+      if (err) return cb(err)
+      const blockStart = newSize - blockSize
+      loadLatestBlock(blockStart, function onTruncateLoadedLatestBlock(err) {
+        if (err) return cb(err)
+        const sizeDiff = size - newSize
+        cb(null, sizeDiff)
+      })
+    })
+  }
+
+  function compact(cb) {
+    if (compaction) {
+      debug('compaction already in progress')
+      waitingCompaction.push(cb)
+      return
+    }
+    onDrain(function startCompactAfterDrain() {
+      compaction = new Compaction(self, (err, sizeDiff) => {
+        compaction = null
+        if (err) return cb(err)
+        compactionSince.set({ value: sizeDiff, done: true })
+        while (waitingCompaction.length) waitingCompaction.shift()()
```

Comment:
convert this to a loop instead

Reply:
Hmm, we already use this `while` pattern in other places, like inside the `raf.stat` that is done when loading the log. So should be consistent

Reply:
Yes agree, I noticed this because we could potentially have a bunch of appends from replication, so the list might be quite large if we block for 10+ s.